src/libstore/build/local-derivation-goal.cc

```diff
@@ -1333,13 +1333,18 @@ struct RestrictedStore : public virtual RestrictedStoreConfig, public virtual Lo
     std::optional<const Realisation> queryRealisation(const DrvOutput & id) override
     // XXX: This should probably be allowed if the realisation corresponds to
     // an allowed derivation
-    { throw Error("queryRealisation"); }
+    {
+        if (!goal.isAllowed(id))
+            throw InvalidPath("cannot query an unknown output id '%s' in recursive Nix", id.to_string());
+        return next->queryRealisation(id);
```

Comment:
Maybe this should be restricted further, to check whether `id` has been built *by this RestrictedStore* (i.e. whether we've seen a `buildPaths`  call to realise this `id`). Otherwise, you could use this to observe whether `id` has previously been built, which would be impure.

Reply:
Maybe the implementation isn’t correct, but my intent is that `isAllowed(DrvOutput & id)` returns true iff
- `id` is a dependency of the current derivation, or
- `id` has been built by the current `RestrictedStore`

(which is the same thing as what’s currently done for `StorePath`s).

Is that wrong?

Reply:
Oh wait, the first condition doesn’t seem to make any sense. I’ll remove that bit

Reply:
OK, I’ve rewritten things so that we don’t track the inputs of the derivation, and `isAllowed(DrvOutput & id)` only returns true if `id` has been built by the current store